### ISSUE TITLE
forEach returns a Null Error for a DataSnapshot but the datasnapshot has no null values in it

### ISSUE THREAD

**Issue by Jasonthefirst** [flutter/flutter#18666]
## Steps to Reproduce

I try to get a DataSnapshot from a firebase database. I get the snapshot. The snapshot looks normal and for 3 similar tables it works without a problem. But for one of the tables I get a problem as soon as it hits forEach. The DataSnapshot looks normal in the debugger and I can view all entries. 

    DataSnapshot snapshot =
        await getSnapshots.getDatasnapshotfromSelection(selection);

    snapshot.value.forEach((d, e) {
      List<String> innerlist = new List();
      var name = e['Name'];
    
      innerlist.add(name);
     [...]
      _nameList.add(innerlist);
    });
## Logs

```
E/flutter ( 3284): type '(dynamic, dynamic) => Null' is not a subtype of type '(dynamic) => dynamic' of 'f'
E/flutter ( 3284): #0      List.forEach (dart:core/runtime/libarray.dart)
E/flutter ( 3284): #1      GetLists.getNameandVideoListFromSelection (package:turn_app/communication/GetLists.dart:51:22)
E/flutter ( 3284): <asynchronous suspension>
E/flutter ( 3284): #2      GetLists.isSelectionLoaded (package:turn_app/communication/GetLists.dart:119:15)
E/flutter ( 3284): <asynchronous suspension>
E/flutter ( 3284): #3      ApparatusButtonStage.checkifIsloaded (package:turn_app/ui/Apparatus_Button.dart:110:42)
E/flutter ( 3284): <asynchronous suspension>
E/flutter ( 3284): #4      ApparatusButtonStage.initState (package:turn_app/ui/Apparatus_Button.dart:135:5)
E/flutter ( 3284): #5      StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:3751:58)
E/flutter ( 3284): #6      ComponentElement.mount (package:flutter/src/widgets/framework.dart:3617:5)
E/flutter ( 3284): #7      Element.inflateWidget (package:flutter/src/widgets/framework.dart:2907:14)
E/flutter ( 3284): #8      Element.updateChild (package:flutter/src/widgets/framework.dart:2710:12)
E/flutter ( 3284): #9      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3653:16)
E/flutter ( 3284): #10     Element.rebuild (package:flutter/src/widgets/framework.dart:3495:5)
E/flutter ( 3284): #11     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3622:5)
E/flutter ( 3284): #12     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3617:5)
E/flutter ( 3284): #13     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2907:14)
E/flutter ( 3284): #14     Element.updateChild (package:flutter/src/widgets/framework.dart:2710:12)
E/flutter ( 3284): #15     SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:4654:14)
E/flutter ( 3284): #16     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2907:14)
E/flutter ( 3284): #17     Element.updateChild (package:flutter/src/widgets/framework.dart:2710:12)
E/flutter ( 3284): #18     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3653:16)
E/flutter ( 3284): #19     Element.rebuild (package:flutter/src/widgets/framework.dart:3495:5)
E/flutter ( 3284): #20     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3622:5)
E/flutter ( 3284): #21     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3617:5)
E/flutter ( 3284): #22     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2907:14)
E/flutter ( 3284): #23     Element.updateChild (package:flutter/src/widgets/framework.dart:2710:12)
E/flutter ( 3284): #24     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3653:16)
E/flutter ( 3284): #25     Element.rebuild (package:flutter/src/widgets/framework.dart:3495:5)
E/flutter ( 3284): #26     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3622:5)
E/flutter ( 3284): #27     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3617:5)
E/flutter ( 3284): #28     ParentDataElement.mount (package:flutter/src/widgets/framework.dart:3955:11)
E/flutter ( 3284): #29     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2907:14)
E/flutter ( 3284): #30     Element.updateChild (package:flutter/src/widgets/framework.dart:2710:12)
E/flutter ( 3284): #31     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3653:16)
E/flutter ( 3284): #32     Element.rebuild (package:flutter/src/widgets/framework.dart:3495:5)
E/flutter ( 3284): #33     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3622:5)
E/flutter ( 3284): #34     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:3769:11)
E/flutter ( 3284): #35     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3617:5)
E/flutter ( 3284): #36     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2907:14)
E/flutter ( 3284): #37     Element.updateChild (package:flutter/src/widgets/framework.dart:2710:12)
E/flutter ( 3284): #38     SliverMultiBoxAdaptorElement.updateChild (package:flutter/src/widgets/sliver.dart:744:36)
E/flutter ( 3284): #39     SliverMultiBoxAdaptorElement.createChild.<anonymous closure> (package:flutter/src/widgets/sliver.dart:729:20)
E/flutter ( 3284): #40     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2222:19)
E/flutter ( 3284): #41     SliverMultiBoxAdaptorElement.createChild (package:flutter/src/widgets/sliver.dart:722:11)
E/flutter ( 3284): #42     RenderSliverMultiBoxAdaptor._createOrObtainChild.<anonymous closure> (package:flutter/src/rendering/sliver_multi_box_adaptor.dart:254:23)
E/flutter ( 3284): #43     RenderObject.invokeLayoutCallback.<anonymous 


```

```
[√] Flutter (Channel dev, v0.5.5, on Microsoft Windows [Version 10.0.17134.112], locale de-DE)
[√] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[√] Android Studio (version 3.1)
[√] IntelliJ IDEA Ultimate Edition (version 2018.1)
[√] Connected devices (1 available)

• No issues found!
```

**Comment by zoechi** [comment]
What's the code of `getSnapshots.getDatasnapshotfromSelection()`?

**Comment by Jasonthefirst** [comment]
```
Future<DataSnapshot> getDatasnapshotfromSelection(Selection selection) async {
    final String reflink = getReflinkFromSelection(selection);
    DatabaseReference reference = 'element/' +
        selection.bundesland +
        '/' +
        selection.gender +
        '/' +
        selection.apparatus;
        globalFBInstanceSingleton.databaseRef.child(reflink);

    DataSnapshot snapshot = await reference.once();
    return snapshot;
  }
```

Is the code for getDataSnapshot. Only Strings (which correspond to tables items)

**Comment by zoechi** [comment]
The problem is that the callback passed to `forEach` only supports one parameter not two 
```dart
snapshot.value.forEach((d, e) {
```
should be
```dart
snapshot.value.forEach((d) {
```

**Comment by Jasonthefirst** [comment]
I don't think this is the case, as I use this for all tables and normally it iterates through the snapshot with the key as d and the value in e. Only in this specific table it breaks. Another table with a different key but the same structure and type of data works without a flaw.

**Comment by Jasonthefirst** [comment]
I seem to have found the error and I think this is kind of a error with the flutter firebase implementation? 
My keys in the first table started with 0 and went further to 20. Because of this the snapshot for the first item with the Id 0 was nested differently. If I removed the item with 0 as the key the first Item with the key 1  was nested differently. This only happened for the first table from 0-20. The second table starting with 21~ didn't have this problem. I was able to change this by uploading my tables with push and letting firebase choose unique keys for me.
